File: src/arli/parse.py
"""Arity-driven parser for arli."""

from __future__ import annotations
from typing import Any, Optional

from .types import Symbol, nil
from .tokenize import (Token, TokenStream, TOKEN_OPEN, TOKEN_CLOSE,
                       TOKEN_VECTOR_OPEN, TOKEN_VECTOR_CLOSE,
                       TOKEN_MAP_OPEN, TOKEN_MAP_CLOSE,
                       TOKEN_NUMBER, TOKEN_STRING, TOKEN_SYMBOL,
                       TOKEN_QUOTE, TOKEN_QUASIQUOTE,
                       TOKEN_UNQUOTE, TOKEN_UNQUOTE_SPLICE,
                       TOKEN_KEYWORD)


class ArityTable:
    """Maps symbol names to their arity. -1 = variadic."""
    def __init__(self) -> None:
        self._table: dict[str, int] = {}
    def register(self, name: str, arity: int) -> None:
        self._table[name] = arity
    def get(self, name: str) -> Optional[int]:
        return self._table.get(name)
    def has(self, name: str) -> bool:
        return name in self._table
    def clone(self) -> ArityTable:
        t = ArityTable(); t._table = dict(self._table); return t


class Parser:
    def __init__(self, arity_table: ArityTable) -> None:
        self.arity_table = arity_table

    def parse(self, source: str) -> list[Any]:
        from .tokenize import tokenize
        tokens = tokenize(source)
        stream = TokenStream(tokens)
        exprs: list[Any] = []
        while not stream.is_eof:
            expr = self._parse_expr(stream, allow_arity=True)
            if expr is not None:
                exprs.append(expr)
        return exprs

    def _parse_expr(self, stream: TokenStream, allow_arity: bool = True) -> Any:
        tok = stream.peek()
        if tok is None:
            return None

        tt = tok[0]
        if tt == TOKEN_OPEN:
            return self._parse_paren_list(stream)
        if tt == TOKEN_VECTOR_OPEN:
            return self._parse_bracket_list(stream)
        if tt == TOKEN_MAP_OPEN:
            return self._parse_map_literal(stream)
        if tt == TOKEN_QUOTE:
            stream.next(); expr = self._parse_expr(stream, False)
            return [Symbol("quote"), expr]
        if tt == TOKEN_QUASIQUOTE:
            stream.next(); expr = self._parse_expr(stream, True)
            return [Symbol("quasiquote"), expr]
        if tt == TOKEN_UNQUOTE:
            stream.next(); expr = self._parse_expr(stream, True)
            return [Symbol("unquote"), expr]
        if tt == TOKEN_UNQUOTE_SPLICE:
            stream.next(); expr = self._parse_expr(stream, True)
            return [Symbol("unquote-splicing"), expr]

        tok = stream.next()
        if tt == TOKEN_NUMBER:
            return tok[1]
        if tt == TOKEN_STRING:
            return tok[1]
        if tt == TOKEN_KEYWORD:
            return Symbol(tok[1])
        if tt == TOKEN_SYMBOL:
            name = tok[1]
            if name in ("defn", "defn-rec"):
                return self._parse_defn_rec(stream)
            if name == "defn-fexpr":
                return self._parse_defn_fexpr(stream)
            if not allow_arity:
                return Symbol(name)
            arity = self.arity_table.get(name)
            if arity is not None and arity >= 0:
                args = []
                for _ in range(arity):
                    arg = self._parse_expr(stream, True)
                    if arg is None:
                        raise SyntaxError(f"Expected {arity} args for {name}, got {len(args)}")
                    args.append(arg)
                return [Symbol(name)] + args
            return Symbol(name)

        raise SyntaxError(f"Unexpected token: {tok}")

    def _parse_paren_list(self, stream: TokenStream) -> list[Any]:
        stream.expect(TOKEN_OPEN)
        items: list[Any] = []
        is_first = True
        # When the first element is quote, disable arity expansion for all
        # subsequent elements (so '(+ 1 2) parses without + consuming args)
        quoted_form = False
        while True:
            tok = stream.peek()
            if tok is None:
                raise SyntaxError("Unclosed parenthesis")
            if tok[0] == TOKEN_CLOSE:
                stream.next(); break
            allow_arity = not is_first and not quoted_form
            expr = self._parse_expr(stream, allow_arity)
            if expr is not None:
                # After parsing the first element, check if it's 'quote'
                if is_first and isinstance(expr, Symbol) and expr.name == 'quote':
                    quoted_form = True
                if (is_first and isinstance(expr, list) and len(expr) > 0
                        and isinstance(expr[0], Symbol) and expr[0].name in ('defn', 'defn-rec', 'defn-fexpr')
                        and stream.peek() is not None and stream.peek()[0] == TOKEN_CLOSE):
                    stream.next(); return expr
                items.append(expr)
            is_first = False
        return items

    def _parse_bracket_list(self, stream: TokenStream) -> list[Any]:
        """Parse [a b c] -> [Symbol('list'), a, b, c]"""
        stream.expect(TOKEN_VECTOR_OPEN)
        items: list[Any] = []
        while True:
            tok = stream.peek()
            if tok is None:
                raise SyntaxError("Unclosed bracket")
            if tok[0] == TOKEN_VECTOR_CLOSE:
                stream.next(); break
            expr = self._parse_expr(stream, True)
            if expr is not None:
                items.append(expr)
        return [Symbol("list")] + items

    def _parse_map_literal(self, stream: TokenStream) -> list[Any]:
        """Parse {:a 1 :b 2} -> [Symbol('hash-map'), :a, 1, :b, 2]"""
        stream.expect(TOKEN_MAP_OPEN)
        items: list[Any] = []
        while True:
            tok = stream.peek()
            if tok is None:
                raise SyntaxError("Unclosed map brace")
            if tok[0] == TOKEN_MAP_CLOSE:
                stream.next(); break
            expr = self._parse_expr(stream, True)
            if expr is not None:
                items.append(expr)
        return [Symbol("hash-map")] + items

    def _parse_name_sym(self, stream: TokenStream) -> Symbol:
        tok = stream.peek()
        if tok is None or tok[0] != TOKEN_SYMBOL:
            raise SyntaxError("Expected a symbol name")
        return Symbol(stream.next()[1])

    def _parse_params_list(self, stream: TokenStream) -> list[Symbol]:
        params = self._parse_expr(stream, False)
        if not isinstance(params, list):
            raise SyntaxError("Expected a parameter list in parens")
        param_syms = []
        for p in params:
            if isinstance(p, Symbol):
                param_syms.append(p)
            else:
                raise SyntaxError(f"Expected symbol in param list, got {p}")
        return param_syms

    def _parse_defn(self, stream: TokenStream) -> list[Any]:
        return self._parse_defn_rec(stream)

    def _parse_fn(self, stream: TokenStream) -> list[Any]:
        param_syms = self._parse_params_list(stream)
        body: list[Any] = []
        while True:
            tok = stream.peek()
            if tok is None or tok[0] in (TOKEN_CLOSE, TOKEN_VECTOR_CLOSE, TOKEN_MAP_CLOSE):
                break
            expr = self._parse_expr(stream, True)
            if expr is not None:
                body.append(expr)
        return [Symbol("fn"), param_syms] + (body or [nil])

    def _parse_defn_rec(self, stream: TokenStream) -> list[Any]:
        name_sym = self._parse_name_sym(stream)
        param_syms = self._parse_params_list(stream)
        self.arity_table.register(name_sym.name, len(param_syms))
        # Parse one body expression (like defn arity 3). Multiple expressions
        # use (do ...) — consistent with defn at top level.
        body: list[Any] = []
        tok = stream.peek()
        if tok is not None and tok[0] not in (TOKEN_CLOSE, TOKEN_VECTOR_CLOSE, TOKEN_MAP_CLOSE):
            expr = self._parse_expr(stream, True)
            if expr is not None:
                body.append(expr)
        return [Symbol("defn"), name_sym, param_syms] + (body or [nil])

    def _parse_defn_fexpr(self, stream: TokenStream) -> list[Any]:
        name_sym = self._parse_name_sym(stream)
        param_syms = self._parse_params_list(stream)
        self.arity_table.register(name_sym.name, len(param_syms))
        body: list[Any] = []
        tok = stream.peek()
        if tok is not None and tok[0] not in (TOKEN_CLOSE, TOKEN_VECTOR_CLOSE, TOKEN_MAP_CLOSE):
            expr = self._parse_expr(stream, True)
            if expr is not None:
                body.append(expr)
        return [Symbol("defn-fexpr"), name_sym, param_syms] + (body or [nil])


def parse_source(source: str, arity_table: Optional[ArityTable] = None) -> list[Any]:
    if arity_table is None:
        arity_table = ArityTable()
    return Parser(arity_table).parse(source)

File: js/arli/src/parse.js
/**
 * Arity-driven parser for arli (JavaScript port).
 */

import { ArliSymbol, nil } from './types.js';
import {
  TokenStream, tokenize, TOKEN_OPEN, TOKEN_CLOSE,
  TOKEN_VECTOR_OPEN, TOKEN_VECTOR_CLOSE,
  TOKEN_MAP_OPEN, TOKEN_MAP_CLOSE,
  TOKEN_NUMBER, TOKEN_STRING, TOKEN_SYMBOL,
  TOKEN_QUOTE, TOKEN_QUASIQUOTE,
  TOKEN_UNQUOTE, TOKEN_UNQUOTE_SPLICE,
  TOKEN_KEYWORD
} from './tokenize.js';

export class ArityTable {
  constructor() { this._table = {}; }
  register(name, arity) { this._table[name] = arity; }
  get(name) { const val = this._table[name]; return val !== undefined ? val : null; }
  has(name) { return Object.prototype.hasOwnProperty.call(this._table, name); }
  clone() { const t = new ArityTable(); t._table = { ...this._table }; return t; }
}

export class Parser {
  constructor(arityTable) { this.arityTable = arityTable; }

  parse(source) {
    const tokens = tokenize(source);
    const stream = new TokenStream(tokens);
    const exprs = [];
    while (!stream.isEOF) {
      const expr = this._parseExpr(stream, true);
      if (expr !== null && expr !== undefined) exprs.push(expr);
    }
    return exprs;
  }

  _parseExpr(stream, allowArity = true) {
    const tok = stream.peek();
    if (tok === null) return null;
    const tt = tok.type;

    if (tt === TOKEN_OPEN) return this._parseParenList(stream);
    if (tt === TOKEN_VECTOR_OPEN) return this._parseBracketList(stream);
    if (tt === TOKEN_MAP_OPEN) return this._parseMapLiteral(stream);
    if (tt === TOKEN_QUOTE) { stream.next(); const expr = this._parseExpr(stream, false); return [new ArliSymbol('quote'), expr]; }
    if (tt === TOKEN_QUASIQUOTE) { stream.next(); const expr = this._parseExpr(stream, true); return [new ArliSymbol('quasiquote'), expr]; }
    if (tt === TOKEN_UNQUOTE) { stream.next(); const expr = this._parseExpr(stream, true); return [new ArliSymbol('unquote'), expr]; }
    if (tt === TOKEN_UNQUOTE_SPLICE) { stream.next(); const expr = this._parseExpr(stream, true); return [new ArliSymbol('unquote-splicing'), expr]; }

    stream.next();
    if (tt === TOKEN_NUMBER) return tok.value;
    if (tt === TOKEN_STRING) return tok.value;
    if (tt === TOKEN_KEYWORD) return new ArliSymbol(tok.value);
    if (tt === TOKEN_SYMBOL) {
      const name = tok.value;
      if (name === 'defn-rec') return this._parseDefnRec(stream);
      if (!allowArity) return new ArliSymbol(name);
      const arity = this.arityTable.get(name);
      if (arity !== null && arity >= 0) {
        const args = [];
        for (let i = 0; i < arity; i++) {
          const arg = this._parseExpr(stream, true);
          if (arg === null || arg === undefined) throw new SyntaxError(`Expected ${arity} args for ${name}, got ${args.length}`);
          args.push(arg);
        }
        return [new ArliSymbol(name)].concat(args);
      }
      return new ArliSymbol(name);
    }
    throw new SyntaxError(`Unexpected token: ${tok.type} (${tok.value})`);
  }

  _parseParenList(stream) {
    stream.expect(TOKEN_OPEN);
    const items = [];
    let isFirst = true;
    let quotedForm = false;
    while (true) {
      const tok = stream.peek();
      if (tok === null) throw new SyntaxError('Unclosed parenthesis');
      if (tok.type === TOKEN_CLOSE) { stream.next(); break; }
      const allowArity = !isFirst && !quotedForm;
      const expr = this._parseExpr(stream, allowArity);
      if (expr !== null && expr !== undefined) {
        if (isFirst && expr instanceof ArliSymbol && expr.name === 'quote') quotedForm = true;
        if (isFirst && Array.isArray(expr) && expr.length > 0 && expr[0] instanceof ArliSymbol && expr[0].name === 'defn' && stream.peek() !== null && stream.peek().type === TOKEN_CLOSE) { stream.next(); return expr; }
        items.push(expr);
      }
      isFirst = false;
    }
    return items;
  }

  _parseBracketList(stream) {
    stream.expect(TOKEN_VECTOR_OPEN);
    const items = [];
    while (true) {
      const tok = stream.peek();
      if (tok === null) throw new SyntaxError('Unclosed bracket');
      if (tok.type === TOKEN_VECTOR_CLOSE) { stream.next(); break; }
      const expr = this._parseExpr(stream, true);
      if (expr !== null && expr !== undefined) items.push(expr);
    }
    return [new ArliSymbol('list')].concat(items);
  }

  _parseMapLiteral(stream) {
    stream.expect(TOKEN_MAP_OPEN);
    const items = [];
    while (true) {
      const tok = stream.peek();
      if (tok === null) throw new SyntaxError('Unclosed map brace');
      if (tok.type === TOKEN_MAP_CLOSE) { stream.next(); break; }
      const expr = this._parseExpr(stream, true);
      if (expr !== null && expr !== undefined) items.push(expr);
    }
    return [new ArliSymbol('hash-map')].concat(items);
  }

  _parseNameSym(stream) {
    const tok = stream.peek();
    if (tok === null || tok.type !== TOKEN_SYMBOL) throw new SyntaxError('Expected a symbol name');
    return new ArliSymbol(stream.next().value);
  }

  _parseParamsList(stream) {
    const params = this._parseExpr(stream, false);
    if (!Array.isArray(params)) throw new SyntaxError('Expected a parameter list in parens');
    const paramSyms = [];
    for (const p of params) {
      if (p instanceof ArliSymbol) paramSyms.push(p);
      else throw new SyntaxError(`Expected symbol in param list, got ${p}`);
    }
    return paramSyms;
  }

  _parseDefn(stream) {
    const nameSym = this._parseNameSym(stream);
    const paramSyms = this._parseParamsList(stream);
    this.arityTable.register(nameSym.name, paramSyms.length);
    const body = [];
    while (true) {
      const tok = stream.peek();
      if (tok === null || tok.type === TOKEN_CLOSE || tok.type === TOKEN_VECTOR_CLOSE || tok.type === TOKEN_MAP_CLOSE) break;
      const expr = this._parseExpr(stream, true);
      if (expr !== null && expr !== undefined) body.push(expr);
    }
    return [new ArliSymbol('defn'), nameSym, paramSyms].concat(body.length > 0 ? body : [nil]);
  }

  _parseFn(stream) {
    const paramSyms = this._parseParamsList(stream);
    const body = [];
    while (true) {
      const tok = stream.peek();
      if (tok === null || tok.type === TOKEN_CLOSE || tok.type === TOKEN_VECTOR_CLOSE || tok.type === TOKEN_MAP_CLOSE) break;
      const expr = this._parseExpr(stream, true);
      if (expr !== null && expr !== undefined) body.push(expr);
    }
    return [new ArliSymbol('fn'), paramSyms].concat(body.length > 0 ? body : [nil]);
  }

  _parseDefnRec(stream) {
    const nameSym = this._parseNameSym(stream);
    const paramSyms = this._parseParamsList(stream);
    this.arityTable.register(nameSym.name, paramSyms.length);
    const body = [];
    const tok = stream.peek();
    if (tok !== null && tok.type !== TOKEN_CLOSE && tok.type !== TOKEN_VECTOR_CLOSE && tok.type !== TOKEN_MAP_CLOSE) {
      const expr = this._parseExpr(stream, true);
      if (expr !== null && expr !== undefined) body.push(expr);
    }
    return [new ArliSymbol('defn'), nameSym, paramSyms].concat(body.length > 0 ? body : [nil]);
  }
}

export function parseSource(source, arityTable = null) {
  if (arityTable === null) arityTable = new ArityTable();
  return new Parser(arityTable).parse(source);
}
